main/src/ca/uwaterloo/flix/language/phase/Inliner.scala

```diff
@@ -38,9 +38,25 @@ object Inliner {
   }
 
   /**
-    * Candidates for inlining are functions with fewer than `InlineThreshold` expressions
+    * A function below the soft threshold is typically inlined.
     */
-  private val InlineThreshold = 8
+  private val SoftInlineThreshold: Int = 4
+
+  /**
+    * A function above the hard threshold is never inlined.
+    */
+  private val HardInlineThreshold: Int = 8
+
+  /**
+    * Returns `true` if `def0` should be inlined.
+    */
+  private def canInlineDef(def0: OccurrenceAst.Def): Boolean = {
+    val mayInline = def0.context.occur != DontInline && !def0.context.isSelfRecursive
+    val belowSoft = def0.context.size < SoftInlineThreshold
+    val belowHard = def0.context.size < HardInlineThreshold
+    val shouldInline = belowSoft || (def0.context.isDirectCall && belowHard) || (def0.context.occur == Once && belowHard)
+    mayInline && shouldInline
```

Comment:
`can = may && should`? :D

Reply:
Sorry what do you mean? Is it a language issue?